PPP: Set parent

0nko opened the pull request:
This PR adds the ability to set the page parent. The screen loads all available pages and and selects the current parent.

Note: This PR depends on wordpress-mobile/WordPress-FluxC-Android#867

0nko commented:
Hey @SylvesterWilmott, @planarvoid suggested that there should be a Save action button in the the Set Parent activity that would explicitly select a page parent. Right now you just tap the back arrow and it could be a bit confusing to users. What do you think?

planarvoid commented on WordPress/src/main/java/org/wordpress/android/viewmodel/pages/PageParentViewModel.kt:
```diff
@@ -3,36 +3,77 @@ package org.wordpress.android.viewmodel.pages
+@Inject constructor(private val pageStore: PageStore, private val resourceProvider: ResourceProvider) : ViewModel() {
+    private val _pages: MutableLiveData<List<PageItem>> = MutableLiveData()
+    val pages: LiveData<List<PageItem>>
+        get() = _pages
```
do you think this approach is more readable? I think functionality wise it's the same.

0nko replied:
You're right. I just used `get()` because it's everywhere else. I'll update other places and get rid of it. It'll save a few lines. Thanks!

SylvesterWilmott commented:
@0nko Yeah for sure, let's go for a save button that becomes visible when a change is made. I'll add it to Zeplin for reference

planarvoid commented:
@0nko the Save button is visible all the time, not just when you make change, otherwise it works as great 👍 
I'd make the click area around the Action button bigger, I keep on clicking next to it and opening the page detail. I think it should match parent in height

SylvesterWilmott commented:
@0nko The save button should have an inactive state with an opacity of 50% and no touch feedback and when changes are made it's at 100% opacity and active. The button style is that of a flat text button ie: 36dp in height with 8dp padding either side and 8dp from the screen edge.

zpl://screen?sid=5b0842b155faa0b37bb1593c&pid=5a86d4456eb267528575ac01
zpl://screen?sid=5b7bce0749b52f4630229ebb&pid=5a86d4456eb267528575ac01

@planarvoid I'm curious as to why you're able to open the page detail from this view? There should be no way to do so especially from the app bar. Maybe I'm missing something

0nko commented on WordPress/src/main/java/org/wordpress/android/ui/pages/PageItemViewHolder.kt:
```diff
@@ -72,15 +78,54 @@ sealed class PageItemViewHolder(internal val parent: ViewGroup, @LayoutRes layou
+
+        private fun selectItem(pageItem: ParentPage) {
+            onParentSelected(pageItem)
+            launch(UI) {
```
Good point 👍.

SylvesterWilmott commented:
It's looking great @0nko !

A few more minor things that I discovered along the way:

1. Pages view: It appears as though the 52dp inset on the tab bar did not respect the tab item paddings. Was this because you accounted for the item paddings or is this how the system handles tab bar paddings? Either way the intended result is to have the text label inside the first tab line up to the 72dp keyline along with the view title. See below:

**Current alignment:**
![current](https://user-images.githubusercontent.com/36532468/44554649-b7908680-a729-11e8-9ab5-1fe6593da45c.png)

**Desired alignment:**
![desired](https://user-images.githubusercontent.com/36532468/44554663-bfe8c180-a729-11e8-974f-e72569e07f3c.png)

2. Pages view: The list cells should be `48dp` in height
3. Select parent: Can we align the text in the cells to the `72dp` keyline here?
4. Select parent: The list cells should be `48dp` in height
5. Select parent: The label `Top level` should be set to the System Font since this option is not user generated content

planarvoid commented:
I managed to crash the app. This is what I did:
1) I created a hierarchy of Page A <- Page B <- Page C where <- means is parent of
2) I moved Page B to draft
3) Page A and Page C are now visible in the flat hierarchy
4) I try to set parent to Page C
5) App crashes - I think Page C still has a parent - page B but it's not in the page list
This is the code that most likely crashes `val parentChoices = pageStore.getPagesFromDb(site).filter { it.remoteId != pageId && it.status == PUBLISHED }` 

```
java.util.NoSuchElementException: Collection contains no element matching the predicate.
        at org.wordpress.android.viewmodel.pages.PageParentViewModel$loadPages$1.doResume(PageParentViewModel.kt:98)
```

SylvesterWilmott commented:
Looks fantastic @0nko !

I guess the empty states are a WIP for now but there are two small issues:

1. Search: Let's not show any helper text/illustration when the input is empty (`Start typing to search`)
2. Search: When searching the clear field `x` on the far right needs some margin from the screen edge

0nko commented:
@theck13 Where did those those empty list designs come from? I don's see them in the PPP Zeplin project. 

@SylvesterWilmott suggested I added actionable empty view state so I just improvised a bit. I removed the blue button because the FAB is always on top of each tab.

SylvesterWilmott commented:
>Where did those those empty list designs come from? I don's see them in the PPP Zeplin project.
Those are in the `Actionable Empty States` Zeplin project

>I removed the blue button because the FAB is always on top of each tab.
The general rule we've been following is to cover the FAB with the empty state (with the blue button included in the empty state)

0nko commented:
I've changed the empty view stat to match the designs. It's ready for another pass.

SylvesterWilmott commented:
Just wanted to mention/re-surface a couple of things @0nko :

1. Search: When searching, the clear field `x` on the far right needs some margin from the screen edge
2. Pages view: The bottom of the list needs some margin (should be 88dp) to account for the FAB. The FAB covers the ellipsis icon on the cells at the bottom of the list.
3. For those empty states with just a string like `What would you like to find?` and `No pages matching your search` They should not be centered vertically. Instead, they should simply have a 56dp top margin

SylvesterWilmott commented:
Just wanted to surface a few things:

1. Search: When searching, the clear field x on the far right needs some margin from the screen edge
2. Pages view: The bottom of the list needs some margin (should be 88dp) to account for the FAB. The FAB covers the ellipsis icon on the cells at the bottom of the list.
3. For those single string empty states `What would you like to find` and `Na pages matching your search` they are not intended to be centered vertically, instead they should just have a top margin of 56dp

SylvesterWilmott commented:
@0nko Just noticed that we have a button on the Trashed Pages tab empty state. We shouldn't show a button on this one.

SylvesterWilmott commented:
@0nko re: the `Fetching pages...` waiting view, can we just remove the illustration from that?

SylvesterWilmott commented:
Another minor thing, can we just change the `What would you like to find` string to `Search pages` instead

0nko commented:
About the title Pages vs Site Pages - the design uses Pages but before it used to be Site Pages. Which one do you perfer @SylvesterWilmott?

0nko commented:
Thanks for the thorough testing, @theck13! All of the issues you've found were edge cases. They should all be fixed now, the PR is ready for another pass.

SylvesterWilmott commented:
@0nko `Site Pages` is fine for the page title.

SylvesterWilmott commented:
>The Trashed tab has the Create floating action button, which I believe should be invisible for that tab like the blue actionable empty view button

That's correct

0nko commented:
I fixed all of the problems mentioned above. Hopefully I didn't miss anything now... 😓

SylvesterWilmott commented:
UI looks great @0nko
